File: backend/app/schemas/analytics.py
from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, ConfigDict


class PauseEvent(BaseModel):
    start_seconds: float
    end_seconds: float
    duration_seconds: float


class WordTimestamp(BaseModel):
    word: str
    start: float
    end: float


class AnalyticsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    interview_id: int
    filler_count: int
    filler_word_breakdown: dict[str, int]
    speaking_rate: float
    pause_count: int
    pauses: List[PauseEvent]
    feedback: str
    overall_score: int


class TranscriptResponse(BaseModel):
    interview_id: int
    full_text: str
    word_timestamps: List[WordTimestamp]


class TimelineEvent(BaseModel):
    type: str
    start_seconds: float
    end_seconds: float
    label: str
    severity: Optional[str] = None


class TimelineResponse(BaseModel):
    interview_id: int
    events: List[TimelineEvent]

File: frontend/src/types/analytics.ts
export interface PauseEvent {
  start_seconds: number;
  end_seconds: number;
  duration_seconds: number;
}

export interface WordTimestamp {
  word: string;
  start: number;
  end: number;
}

export interface AnalyticsResult {
  interview_id: number;
  filler_count: number;
  filler_word_breakdown: Record<string, number>;
  speaking_rate: number;
  pause_count: number;
  pauses: PauseEvent[];
  feedback: string;
  overall_score: number;
}

export interface TranscriptResponse {
  interview_id: number;
  full_text: string;
  word_timestamps: WordTimestamp[];
}

export interface TimelineEvent {
  type: string;
  start_seconds: number;
  end_seconds: number;
  label: string;
  severity?: string | null;
}

export interface TimelineResponse {
  interview_id: number;
  events: TimelineEvent[];
}

export interface ProgressPoint {
  interview_id: number;
  title: string;
  created_at: string;
  overall_score: number;
  speaking_rate: number;
  filler_count: number;
}

export interface UserProgress {
  total_interviews: number;
  completed_interviews: number;
  average_score: number;
  average_speaking_rate: number;
  average_filler_count: number;
  history: ProgressPoint[];
}
